src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Infrastructure/HttpUriHelper.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    internal class HttpUriHelper : UriHelperBase
+    {
+        private readonly HttpContext _context;
+
+        public HttpUriHelper(HttpContext context)
+        {
+            _context = context;
+        }
+
+        protected override void InitializeState()
+        {
+            SetAbsoluteBaseUri(GetContextBaseUri());
+            SetAbsoluteUri(GetFullUri());
+        }
+
+        private string GetFullUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(
+                request.Scheme,
+                request.Host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+        }
+
+        private string GetContextBaseUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase);
+        }
+
+        protected override void NavigateToCore(string uri, bool forceLoad)
+        {
+            // For now throw.
```

Comment:
I'm specifically avoiding dealing with redirects as of right now so that we can discuss how to better handle this post preview3.

Reply:
> The bigger (not easily solvable problem) is that the response might have already
> started (for example when the page/view has already been flushed)

Is that really possible? Am I right in thinking it's only possible *after* the initial async render? I know that components can include arbitrary async logic that runs on a separate thread (or non-awaited continuation) and tries to trigger a redirect after the page has been rendered. But if we impose the rule "we only listen to your attempts to navigate while the initial async render is in progress", we can be sure it hasn't started writing to the response, right? Because we don't even start transforming the rendertree into HTML until the initial async render is done.

Reply:
But this can happen inside an MVC application where you could have already called FlushAsync to start sending the head to the page (to speed-up script loading) for example.

We might be Ok by checking response.started or something similar, but I want to think about it before baking it in

Reply:
Oh, that scenario sounds fine. The rules around when you can issue redirections in MVC are well established (i.e., not after you've started flushing output), so this wouldn't be any different. As long as it only fails when the developer has explicitly taken steps to start flushing the output, I think we're fine.